[FEATURE] Q&A 도메인 및 사용자 측 기능 구현

JakePark929 opened the pull request:
## 📝 PR Summary

Q&A 도메인을 추가하고 사용자 측 기능을 구현 완료 했습니다.
자세한 개발 내용은 아래와 같습니다.

- `Qna` 도메인이 추가되었습니다. 유저와 관리자 아이디를 `User` 도메인과 각각 N:1 단방향 맵핑 하도록 하였습니다.
- Q&A의 사용자측 기능인 내 Q&A 목록 조회, Q&A 세부내용 조회, Q&A 등록이 추가되었습니다.
   - Q&A 목록 조회 : 프로젝션을 이용하여 필요한 정보만 취득하고 페이징처리 하도록 구현했습니다.
   - Q&A 세부조회 : 우선 Q&A 전체 내용을 응답 받도록 구현했습니다.
   - Q&A 등록 : title, content 를 전달 받으면 새로운 Qna 가 등록됩니다.

#### 🌲 Working Branch

feat/#16-qna-client

#### 🌲 TODOs

- [x] Q&A 추가
- [x] 내 Q&A 목록 조회
- [x] Q&A 세부 조회

### Related Issues

resolve #16

JakePark929 committed:
[#16] feat : QnA 도메인 및 사용자 기능 구현

- `Qna` 도메인이 추가되었습니다. 유저와 관리자 아이디를 `User` 도메인과 각각 1:N 단방향 맵핑 하도록 하였습니다.
- Q&A의 사용자측 기능인 내 Q&A 목록 조회, Q&A 세부내용 조회, Q&A 등록이 추가되었습니다.
> Q&A 목록 조회는 프로젝션을 이용하여 필요한 정보만 취득하고 페이징처리 하도록 구현했습니다.

donsonioc2010 commented on storage/db-core/src/main/java/com/nerd/favorite18/storage/db/core/qna/entity/Qna.java:
```diff
@@ -0,0 +1,62 @@
+@Table(name = "tbl_qna")
+@Getter
+@Setter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
```
Entity에서 Class위에 `@Comment("코멘트 명칭")`을 달면 테이블에,
Column 필드 에다가 `@Comment("컬럼 코멘트")`를 달아두면, Auto로 테이블이 생성시에 코멘트를 달아주더라구요. 

한번 저희도 해보는거 어떨까요 ㅋㅋ

Ps. 참고로 테이블에 적용되는건 SpringBoot3에서 부터입니다. 2 이전에는 컬럼만 지원합니다.

JakePark929 replied:
코멘트 한번 추가 해봤는데 최신 버전으로 주석 넣은 건 처음이라 더 나은 컨벤션이 있으면 제안 부탁드립니다~

donsonioc2010 commented on storage/db-core/src/main/java/com/nerd/favorite18/storage/db/core/qna/projection/QnaListResponse.java:
```diff
@@ -0,0 +1,15 @@
+    QnaProgressStatus getProgressStatus();
+    AnswerStatus getAnswerStatus();
+    LocalDateTime getCreatedAt();
+    LocalDateTime getUpdatedAt();
```
getId와 함께, 해당 두개는 Interface로 상위에 한번더 추출해두는게 가능하지 않을까요?

JakePark929 replied:
id, createdAt, updatedAt 에 대해서 BaseProjection 이라는 인터페이스로 빼봤습니다 :)

donsonioc2010 commented on storage/db-core/src/main/java/com/nerd/favorite18/storage/db/core/qna/projection/QnaListResponse.java:
```diff
@@ -0,0 +1,15 @@
+
+import java.time.LocalDateTime;
+
+public interface QnaListResponse {
```
해당 Interface의 경우 어떤 의도로 사용을 하고자 했던 부분인지 궁금합니다. :)

처음보는 방식이어서요 (싫다 좋다가 아니라 진짜 몰라서입니다.)

JakePark929 replied:
아래 Jpa Repository 에서 보셔서 아시겠지만 JPA를 사용하여 DB를 조회하는 경우, 
특히 원하는 컬럼만 조회하고 싶은 경우 여러가지 방식이 있다고 알고 있습니다.

QueryDSL 라이브러리를 쓰는 것도 물론 좋겠지만, 
그렇지 못한 경우 네이티브 쿼리, JPQL 그리고 이번에 제가 사용한 인터페이스를 통한 프로젝션 방식입니다.
이 외에도 entityManager를 직접 구현하는 등 여러 방식이 있던 걸로 기억합니다.

페이지 처리 같은 경우는 Page 객체에 엔티티 객체를 제네릭으로 넣어줘야만  처리가 되는데
엔티티를 컨트롤러 까지 가져가서 반환해야 되나라는 생각에 Dto로 페이징 처리하려 했구요.

저도 사실 이런 인터페이스 프로젝션 방식은 이번에 처음 써봤는데
원하는 컬럼만 간단하게 가져올 때 쓰기 좋은 방법인 것 같습니다!
요구사항 정의서 보니 Q&A 리스트는 제목만 가져오면 되더라구요~

아래 링크는 Spring JPA 에 Projection 에 관한 공식 문서입니다 :D
https://docs.spring.io/spring-data/jpa/reference/repositories/projections.html

donsonioc2010 commented on storage/db-core/src/main/java/com/nerd/favorite18/storage/db/core/qna/repository/QnaRepository.java:
```diff
@@ -0,0 +1,14 @@
+
+
+public interface QnaRepository extends JpaRepository<Qna, Long> {
+    Page<QnaListResponse> findAllByQnaUserOrderByCreatedAtDesc(User qnaUser, Pageable pageable);
```
오호 이런식으로 사용이 가능하군요.. QNA가 상속을 받는게 아닌데 어떻게 가능한건가요??

혹시 제가 볼수 있는 자료같은거 링크좀 요청드려도 괜찮을까요?

JakePark929 replied:
위 코멘트에 설명 드렸습니다~ :)

donsonioc2010 reviewed (approved):
여쭤보고 싶은부분만 코멘트 달아놨습니당 :)

JakePark929 committed:
[#16] refactor : 코멘트 어노테이션 추가 프로젝션 등 리뷰내용 반영

- `Qna` 와 `BaseEntity`에 코멘트를 추가했습니다.
- response 용 인터페이스의 id, createdAt, updatedAt 을 분리하여 상위 인터페이스인 `BaseProjection`에서 상속받도록 했습니다.

birdieHyun commented on storage/db-core/src/main/java/com/nerd/favorite18/storage/db/core/qna/repository/QnaRepository.java:
```diff
@@ -0,0 +1,14 @@
+
+public interface QnaRepository extends JpaRepository<Qna, Long> {
+    Page<QnaListResponse> findAllByQnaUserOrderByCreatedAtDesc(User qnaUser, Pageable pageable);
+    Qna findByIdAndQnaUserOrderByIdDesc(Long id, User qnaUser);
```
Qna 객체가 null 일 가능성은 없을까요?

JakePark929 replied:
Qna null 체크를 놓쳤네요!
Optional 객체 추가하여 처리 해줬습니다.

덕분에 놓쳤던 부분까지 확인 했습니다. 감사합니다 :)

birdieHyun commented on storage/db-core/src/main/java/com/nerd/favorite18/storage/db/core/qna/entity/Qna.java:
```diff
@@ -0,0 +1,72 @@
+    @ManyToOne(fetch = FetchType.LAZY)
+    private User adminUser;
+
+    @Comment("답변 내용")
+    @Column
+    @Lob
+    private String answerContent;
```
@Lob 어노테이션은 제가 사용해보지 않아서 정확하진 않을수 있어서 참고만 부탁드릴게요! 

Lob을 사용하게 되면 mysql 내부에서 blob(혹은 clob) 과 같은 대용량 데이터를 사용하는 형태로 저장될 것 같아요
answerContent의 경우 그렇게 큰 데이터를 저장할것으로 보이지 않는데 varchar가 아닌 다른 형태로 저장하신 이유가 있으신가요?

JakePark929 replied:
아, 이 부분은 저도 코드 작성하면서 Q&A 질문, 답변 컬럼에 왜 CLOB 을 쓸까 생각했는데,
ERD에 형식이 CLOB 으로 되어있어서 우선 Lob 을 추가하였습니다.
Q&A 쪽 설계 담당했던 종원 님에게 확인 해보겠습니다 :D

donsonioc2010 replied:
CLob또는 Text를 염두해 두고 있었는데요, Varchar를 대신하여 사용한 이유는 아래와 같은 단계형식으로 생각이 진화했습니다.

- Content에 대해서는, 많은 텍스트를 작성하게 될 가능성이 높은 부분이라 생각
    - 그러면 Varchar가 저장하는 4000으로는 적을 수 있겠다
        - 한글이 3바이트였는지, 2바이트였는지 약간 혼돈은 되는데, 2바이트라 쳐도 2000천자 정도구나 (3바이트였던걸로 기억은 납니다)
        - 관리자는 웹으로 구현할 가능성이 높기에 웹 에디터를 사용할 가능성도 높겠다. 
            - CLob 이 더 적당하겠다.

다음의 과정으로 판단하였습니다.

birdieHyun commented on storage/db-core/src/main/java/com/nerd/favorite18/storage/db/core/BaseEntity.java:
```diff
@@ -19,11 +20,11 @@ public abstract class BaseEntity {
     private LocalDateTime createdAt;
 
     @UpdateTimestamp
-    @Column
+    @Column @Comment("수정 일자")
     private LocalDateTime updatedAt;
```
@Column 이 어떤 역할을 해주는건가요?

JakePark929 replied:
https://hibernate.atlassian.net/issues/HHH-4369?jql=project%20%3D%20%22HHH%22%20AND%20fixVersion%20%3D%20%225.6.0.Beta2%22

위 링크 코멘트 어노테이션 관련 문서를 보고 @Column 어노테이션이 있어야 @Comment 가 동작하는 건가 싶어서 추가했었는데 방금 확인해보니 없어도 잘 동작 합니다! 문서가 읽기 어렵게 되어있어서 잘못 이해한 듯 합니다.
확인 감사합니다~

birdieHyun commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/qna/service/QnaService.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    public QnaDto getQna(UserDto userDto, Long id) {
+        User userEntity = Optional.ofNullable(
+                userRepository.findFirstByIdAndStatusOrderByIdDesc(userDto.getId(), UserStatus.ACTIVE)
+        ).orElseThrow(() -> new CoreApiException(ErrorType.USER_NOT_FOUND));
```
null 처리에 대해 고민하신게 보여서 좋네요! 
하지만 repository 코드를 직접 작성하지 않은 사용자라면 user entity가 null로 오지 않을 것이라고 예상할 수도 있을 것 같은데,
이러한 부분은 어떻게 예방할 수 있을까요?

JakePark929 replied:
User 도 마찬가지로 Optional 로 처리하도록 하겠습니다!
이 쪽 브랜치는 Qna 관련 브랜치여서 회원 탈퇴 쪽 PR에서 수정 하겠습니다~

birdieHyun commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/qna/dto/QnaAddRequest.java:
```diff
@@ -0,0 +1,13 @@
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class QnaAddRequest {
+    private String title;
+    private String content;
+}
```
dto라서 크게 상관은 없지만 Data 어노테이션을 사용하신 이유가 있으신가요? 
또한 NoArgsConstructor와 AllArgsConstructor 경우 어디서 어떻게 사용되는걸까요?

JakePark929 replied:
디버깅할 때 가끔 쓰다보니 습관적으로 추가했는데요!
확인 해보니 @Getter 만 있으면 될 것 같습니다. 감사합니다~

birdieHyun commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/qna/business/QnaBusiness.java:
```diff
@@ -0,0 +1,33 @@
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+
+@RequiredArgsConstructor
+@Business
+public class QnaBusiness {
```
Business 어노테이션의 경우 어떤 역할을 해주는걸까요?

JakePark929 replied:
Business 어노테이션은 제가 Custom 으로 만든 어노테이션인데요~
내부 동작은 Service 어노테이션과 동일 하지만 Service 클래스와 구별하기 위해 추가했습니다!

서비스 클래스와 비지니스 클래스 역할은 아래와 같습니다.
- 서비스 클래스 : 엔티티에 관한 작업만 수행하며, 리포지터리에 엔티티 전달 또는 비즈니스 클래스에 dto를 반환합니다. 이로서 엔티티의 불필요한 외부 전파를 막습니다.
- 비즈니스 클래스 : 그 외 도메인에 대한 작업을 수행하고 서비스와 dto를 주고 받는 클래스 입니다.

#7 PR 에서 종원님도 같은 질문해주셨는데 유저, 인증 기능 구현한 부분도 한 번 리뷰 해주시고, 의견 주시면 감사하겠습니다~~

JakePark929 committed:
[#16] refactor : Qna null 체크 외 리뷰사항 반영

- 리포지터리에 Optional 객체를 이용해 null 체크를 진행했습니다.
- 불필요한 @Column 어노테이션을 제외했습니다.
- dto 에서 불필요한 어노테이션을 제거했습니다.

JakePark929 committed:
[#16] refactor : User Optional null 체크 추가

- User 리포지터리에 Optional 객체를 이용해 null 체크를 진행했습니다.

JakePark929 commented:
다들 리뷰 열심히 참여 해주셔서 감사합니다. 본 pr 머지 완료 했습니다:)
